### PR TITLE
fix(web-shell): polish session timeline rail

### PR DESCRIPTION
**Pull request by callmeYe**
## What this PR does

Polishes the session timeline rail in the web shell so it behaves like a compact navigation aid instead of competing with the conversation content. The rail now uses tighter tick sizing, sits closer to the message column, hides in wide chat mode, drops the browser-native hover title, only appears when there is enough timeline density, and shows cleaned summaries that match the visible assistant reply more closely.

## Why it's needed

The merged timeline UI exposed a few visual regressions: the rail was offset too far right, hover affordances were oversized, the native title overlapped the custom details card, sparse sessions showed unnecessary chrome, and markdown markers or thinking text could make the details card feel disconnected from the actual answer on screen.

### How to verify

Open a session with four or more assistant entries and confirm the left timeline appears as a compact rail with the longest active/hovered tick around 38px while the shortest tick remains unchanged. Hover or focus a tick and confirm the custom details card is close to the rail, no browser-native title appears, and markdown emphasis markers are removed from the preview text. Switch the chat width to wide mode and confirm the rail hides automatically. Open a session with fewer than four entries and confirm the rail is not rendered.

### Evidence (Before & After)

Before: screenshots showed the rail too far from the message column, oversized active ticks, a native title over the custom details card, and previews containing raw markdown/thinking text. After: the rail is shifted left, ticks are visually toned down, the details card is closer and cleaner, wide mode suppresses the rail, and sparse sessions do not show it.

### Tested on

|     OS     | Status |
| :--------: | :----: |
|  🍏 macOS  | ✅ tested |
| 🪟 Windows | ⚠️ not tested |
|  🐧 Linux  | ⚠️ not tested |

### Environment (optional)

Local web-shell validation: `cd packages/web-shell && npx vitest run client/components/MessageList.test.ts client/components/MessageList.dom.test.tsx` passed with 102 tests; `cd packages/web-shell && npm run build` passed with only the existing Vite chunk-size warning.

## Risk & Scope

- Main risk or tradeoff: This is intentionally scoped to the session timeline presentation and summary generation; hover sizing is tuned visually rather than adding new configuration.
- Not validated / out of scope: Full cross-platform browser visual QA beyond local macOS validation.
- Breaking changes / migration notes: None.

## Linked Issues

N/A

<details>
<summary>中文说明</summary>

## What this PR does

本 PR 优化 web shell 里的 session timeline rail，让它更像轻量导航辅助，而不是和对话内容抢视觉焦点。调整后 rail 更靠近消息区域，横线尺寸更收敛，宽屏聊天模式下自动隐藏，不再触发浏览器原生 title，只有足够多的 entry 时才展示，并且详情卡片里的摘要会清理 markdown 标记，内容也更贴近页面上实际可见的助手回复。

## Why it's needed

合入后的 timeline UI 暴露了几个视觉问题：rail 位置偏右，hover 状态横线过长，原生 title 会压在自定义详情卡片上，entry 很少的会话也显示了不必要的 rail，并且 markdown 标记或 thinking 文本会让详情卡片内容看起来和页面回复不一致。

### How to verify

打开一个包含至少四个 assistant entry 的 session，确认左侧 timeline 会展示为紧凑 rail，active/hover 时最长横线约 38px，最短横线保持不变。hover 或 focus 某条横线时，确认自定义详情卡片离 rail 更近，没有浏览器原生 title，并且预览文本里不会露出 markdown 强调标记。切换到宽屏聊天模式，确认 rail 自动隐藏。打开少于四个 entry 的 session，确认 rail 不渲染。

### Evidence (Before & After)

Before: 截图里 rail 离消息区域偏远，active 横线过长，原生 title 覆盖自定义详情卡片，预览里会露出 raw markdown 或 thinking 文本。After: rail 左移、横线视觉收敛、详情卡片更近且更干净，宽屏模式隐藏 rail，稀疏 session 不再展示 rail。

### Tested on

|     OS     | Status |
| :--------: | :----: |
|  🍏 macOS  | ✅ tested |
| 🪟 Windows | ⚠️ not tested |
|  🐧 Linux  | ⚠️ not tested |

### Environment (optional)

本地 web-shell 验证：`cd packages/web-shell && npx vitest run client/components/MessageList.test.ts client/components/MessageList.dom.test.tsx` 通过，共 102 个测试；`cd packages/web-shell && npm run build` 通过，仅有既有的 Vite chunk-size warning。

## Risk & Scope

- Main risk or tradeoff: 本 PR 有意只覆盖 session timeline 的展示和摘要生成；hover 尺寸基于视觉协调调整，没有新增配置。
- Not validated / out of scope: 除本地 macOS 外，没有做完整跨平台浏览器视觉 QA。
- Breaking changes / migration notes: 无。

## Linked Issues

N/A

</details>

### CODE REVIEW

**Review by ytahdn — COMMENTED** [review]
## Code review summary

**Direction:** ✅ Looks good — Tight visual polish of the session timeline rail. Each change maps to a specific regression listed in the motivation. Scope is right.

**Approach:** ✅ Looks good — `hideSessionTimeline` prop, min-entry gating, regex-based markdown cleaning (avoids pulling `remark` + `strip-markdown` into web-shell), and final-answer-preference in `timelineDetailForTurn` all pull their weight.

**Code review:** ✅ Looks good — Build passes, 110 tests pass (86 unit + 24 DOM). Two minor observations below.

---

## Findings

### [minor] `cleanTimelineMarkdown` italic regex misses non-ASCII punctuation

The italic lookahead `[.,;:!?，。；：！？]` does not cover all Unicode punctuation categories. A CJK user message like `这是 *强调* 内容？` where the trailing punctuation is a Unicode category member not in the explicit list (e.g. `·`, `—`, `「`) would leave the asterisks in the output.

Could use a Unicode-aware alternative:
```ts
.replace(/(^|\s)\*([^*\s][^*]*?\S)\*(?=\s|$|\p{P})/gu, '$1$2')
```

Low-impact since `**bold**` is far more common than `*italic*` in assistant output, but worth noting for CJK-heavy content.

### [minor] Observer churn when timeline visibility props change

The `useLayoutEffect` at line ~1783 creates/destroys the `ResizeObserver` whenever `hideSessionTimeline` or `hasEnoughSessionTimelineEntries` flips. In practice this is rare (chat-width mode toggle, or crossing the 4-entry threshold), but each flip tears down and rebuilds the observer — which can briefly miss resize events during the transition.

Could keep the observer always alive and check the gate inside the callback:
```ts
const updateVisibility = () => {
  if (hideSessionTimeline || !hasEnoughSessionTimelineEntries) {
    setIsSessionTimelineVisible(prev => (prev ? false : prev));
    return;
  }
  const width = el.getBoundingClientRect().width;
  const nextVisible = width >= 1160;
  setIsSessionTimelineVisible(prev => prev === nextVisible ? prev : nextVisible);
};
```

---

## What's good

- Extracting the magic number `30` into `FOLLOW_BOTTOM_THRESHOLD_PX` improves readability and makes the threshold discoverable.
- `reportCanScrollToBottom` with its `scheduleScrollOverflowReport` debounce avoids over-reporting during rapid layout changes (streaming, disclosure expansion).
- `handleDisclosureClickCapture` on the root div via capture phase elegantly covers thinking-block, tool-result, and code-block disclosures without per-component wiring.
- `timelineDetailForTurn` preferring the visible final answer over the node-kind summary is the right call — it makes the timeline detail card match what the user actually sees on screen.
- The `setShouldFollow` early-return when the value is unchanged is a clean optimization that also simplifies the reporting logic.
- Test coverage is thorough: 10 new scroll-to-bottom tests, 4 timeline-visibility tests, markdown-cleaning test, and the `simpleTurns` helper reduces boilerplate in existing tests.

File: packages/web-shell/client/components/MessageList.tsx [hunk 1]
```diff
@@ -541,6 +542,20 @@ function compactTimelineText(
...
 
+function cleanTimelineMarkdown(raw: string | null | undefined): string {
+  if (!raw) return '';
+  return raw
+    .replace(/!\[([^\]]*)\]\([^)]+\)/g, '$1')
+    .replace(/\[([^\]]+)\]\([^)]+\)/g, '$1')
+    .replace(/`([^`]+)`/g, '$1')
+    .replace(/~~([^~]+)~~/g, '$1')
+    .replace(/(\*\*|__)(?=\S)([\s\S]*?\S)\1/g, '$2')
+    .replace(/(^|\s)\*([^*\s][^*]*?\S)\*(?=\s|$|[.,;:!?，。；：！？])/g, '$1$2')
```
Comment (ytahdn) [inline packages/web-shell/client/components/MessageList.tsx]: **[minor]** The italic lookahead `[.,;:!?，。；：！？]` does not cover all Unicode punctuation categories. A user message like `这是 *强调* 内容·` (with `·`, `—`, `「`, etc.) would leave the asterisks in the output.

Could use a Unicode-aware alternative:
```ts
.replace(/(^|\s)\*([^*\s][^*]*?\S)\*(?=\s|$|\p{P})/gu, '$1$2')
```

Low-impact since `**bold**` is far more common than `*italic*` in assistant output, but worth noting for CJK-heavy content.

File: packages/web-shell/client/components/MessageList.tsx [hunk 2]
```diff
@@ -1751,8 +1775,15 @@ export const MessageList = forwardRef<MessageListHandle, MessageListProps>(
 
     const [isSessionTimelineVisible, setIsSessionTimelineVisible] =
       useState(false);
+    const hasEnoughSessionTimelineEntries =
+      sessionTimelineEntries.length >= SESSION_TIMELINE_MIN_VISIBLE_ENTRIES;
 
     useLayoutEffect(() => {
```
Comment (ytahdn) [inline packages/web-shell/client/components/MessageList.tsx:1832]: **[minor]** This `useLayoutEffect` creates/destroys the `ResizeObserver` whenever `hideSessionTimeline` or `hasEnoughSessionTimelineEntries` flips. Each flip tears down and rebuilds the observer, which can briefly miss resize events during the transition.

Could keep the observer always alive and check the gate inside the callback:
```ts
const updateVisibility = () => {
  if (hideSessionTimeline || !hasEnoughSessionTimelineEntries) {
    setIsSessionTimelineVisible(prev => (prev ? false : prev));
    return;
  }
  const width = el.getBoundingClientRect().width;
  const nextVisible = width >= 1160;
  setIsSessionTimelineVisible(prev => prev === nextVisible ? prev : nextVisible);
};
```

In practice this is rare (chat-width mode toggle or crossing the 4-entry threshold), so low risk.

File: packages/web-shell/client/components/MessageList.tsx [hunk 3]
```diff
@@ -541,6 +542,20 @@ function compactTimelineText(
...
+  if (!raw) return '';
+  return raw
+    .replace(/!\[([^\]]*)\]\([^)]+\)/g, '$1')
+    .replace(/\[([^\]]+)\]\([^)]+\)/g, '$1')
+    .replace(/`([^`]+)`/g, '$1')
+    .replace(/~~([^~]+)~~/g, '$1')
+    .replace(/(\*\*|__)(?=\S)([\s\S]*?\S)\1/g, '$2')
+    .replace(/(^|[\s\p{P}])\*([^*\s][^*]*?\S)\*(?=\s|$|\p{P})/gu, '$1$2')
+    .replace(/^\s{0,3}#{1,6}\s+/gm, '')
+    .replace(/^\s{0,3}>\s?/gm, '')
```
Comment (ytahdn) [inline packages/web-shell/client/components/MessageList.tsx:565]: **[Suggestion]** The italic lookbehind `(^|[\s\p{P}])` does not cover CJK characters adjacent to `*italic*`. For example, `好*强调*` will not match because `好` is neither `\s` nor `\p{P}` — it's a `\p{L}` (Letter) character. This is a high-frequency pattern in CJK user messages where there are typically no spaces between characters.

Similarly, `_italic_` is not handled (bold covers both `**` and `__`, but italic only covers `*`).

Consider broadening the lookbehind/lookahead:
```ts
.replace(/(^|[^\p{L}\p{N}])\*([^*\s][^*]*?\S)\*(?=[^\p{L}\p{N}]|$)/gu, '$1$2')
```

This matches italic markers preceded/followed by any non-letter, non-number character (or string boundary), which covers CJK chars, symbols, and punctuation uniformly.

_— qwen3.7-max via Qwen Code /review_

**Review by DragonnZhang — COMMENTED** [review]
⚠️ Downgraded from Approve to Comment: CI still running.

This PR polishes the session timeline rail in web-shell: hides it in wide mode, requires at least 4 turns before showing it, strips markdown from timeline labels/details, and prefers the final answer text in the detail tooltip. CSS tweaks adjust tick sizing and positioning. Changes are clean, well-tested, and correct.

_— qwen3-coder via Qwen Code /review_

File: packages/web-shell/client/components/MessageList.tsx [hunk 4]
```diff
@@ -605,6 +624,18 @@ function timelineDetailForTurn(
   finalAssistantId: string | null,
   nodeKinds: readonly TurnTimelineNodeKind[],
 ): string {
+  if (finalAssistantId !== null) {
```
Comment (wenshao) [inline packages/web-shell/client/components/MessageList.tsx:669]: **[Critical]** This early return can treat a pre-tool assistant update as the turn's final answer. `getSessionTimelineEntries()` currently picks `finalAssistantId` by scanning for the last non-streaming assistant anywhere in the turn, while the collapse path uses `findFinalAnswerIndex()` to require the answer to appear after the last tool/plan work step. For a turn like `user -> assistant("I'll check") -> tool_group`, the hover detail now returns `I'll check` and suppresses `1 tool call`, so the timeline rail disagrees with the visible collapsed transcript.

Please select the preferred final answer with the same after-last-work-step rule used by collapse, or only take this branch when the assistant candidate is after the last tool/plan/parallel-agent item. A regression test for assistant-before-tool with no later final answer would pin this down.

_— GPT-5 via Qwen Code /review_

File: packages/web-shell/client/components/MessageList.tsx [hunk 5]
```diff
@@ -532,7 +533,7 @@ function compactTimelineText(
   raw: string | null | undefined,
   maxLength: number,
 ): string {
-  const compact = raw?.replace(/\s+/g, ' ').trim() ?? '';
+  const compact = cleanTimelineMarkdown(raw).replace(/\s+/g, ' ').trim();
```
Comment (wenshao) [inline packages/web-shell/client/components/MessageList.tsx]: **[Suggestion]** `compactTimelineText()` now strips markdown for every caller, including prompt and `user_shell` timeline labels. Shell commands are not markdown, so labels can show a different command than the one that actually ran: `find packages/*/src/*.ts` becomes `find packages//src/.ts`, and backtick commands lose their backticks.

Please keep raw compaction as the default for command/prompt labels and opt into `cleanTimelineMarkdown()` only for markdown-bearing assistant/detail text, or split this into separate helpers for labels versus assistant summaries.

_— GPT-5 via Qwen Code /review_

**Review by wenshao — CHANGES_REQUESTED** [review]

File: packages/web-shell/client/components/MessageList.tsx (hunk 4 above)
Comment (callmeYe) [reply packages/web-shell/client/components/MessageList.tsx:669]: Fixed in `7dbd6d969`: timeline entries now choose the preferred final assistant answer with the same after-last-work-step rule as collapse, using the grouped turn items and a virtual turn-start boundary. A `user -> assistant("I'll check") -> tool_group` regression now keeps the tool-call detail visible instead of treating the pre-tool assistant text as final.

Also fixed the related label issue in the same commit: raw compaction is the default for prompt/shell labels, and markdown cleanup is opt-in for assistant/detail text. Added shell glob coverage so `find packages/*/src/*.ts` stays unchanged.

Validation run locally:
- `cd packages/web-shell && npx vitest run client/components/MessageList.test.ts client/components/MessageList.dom.test.tsx` - 112 passed
- `cd packages/web-shell && npm run build` - passed

**Review by DragonnZhang — DISMISSED** [review]
**APPROVE** — Incremental changes since f28d0f37 look correct.

The `hideSessionTimeline` prop flows cleanly from App to MessageList, the 4-entry minimum threshold is sensible, `cleanTimelineMarkdown` handles the common formatting cases well, and `timelineDetailForTurn` correctly prioritizes the final answer snippet. The `useLayoutEffect` dependency array is properly updated, and the test suite is thorough with good coverage of the new visibility conditions.

_— qwen3-coder via Qwen Code /review_

**Comment by wenshao** [comment]
@qwen-code /review

File: packages/web-shell/client/components/MessageList.tsx [hunk 6]
```diff
@@ -541,6 +545,24 @@ function compactTimelineText(
...
+function cleanTimelineMarkdown(raw: string | null | undefined): string {
+  if (!raw) return '';
+  return raw
+    .replace(/!\[([^\]]*)\]\([^)]+\)/g, '$1')
+    .replace(/\[([^\]]+)\]\([^)]+\)/g, '$1')
+    .replace(/`([^`]+)`/g, '$1')
+    .replace(/~~([^~]+)~~/g, '$1')
+    .replace(/(\*\*|__)(?=\S)([\s\S]*?\S)\1/g, '$2')
+    .replace(/\*([^*\s][^*]*?\S)\*/g, '$1')
+    .replace(
```
Comment (wenshao) [inline packages/web-shell/client/components/MessageList.tsx:573]: **[Critical]** The bold regex `/(\*\*|__)(?=\S)([\s\S]*?\S)\1/g` has O(n²) backtracking on input with unclosed `**` sequences. The lazy `[\s\S]*?` scans past subsequent `**` boundaries because the backreference `\1` prevents the engine from short-circuiting.

**Impact:** A crafted assistant response with repeated `**` tokens without matching closers can freeze the browser tab. The function re-runs during streaming (signature-cached but invalidated each token), so the cost compounds as content grows.

```suggestion
    .replace(/\*\*(?=\S)([^*]|\*(?!\*))+?(?<=\S)\*\*/g, '$1')
    .replace(/__(?=\S)([^_]|_(?!_))+?(?<=\S)__/g, '$1')
```

_— qwen3.7-max via Qwen Code /review_

File: packages/web-shell/client/components/MessageList.tsx [hunk 7]
```diff
@@ -605,6 +627,20 @@ function timelineDetailForTurn(
...
+    for (const item of turnItems) {
+      if (item.type !== 'message') continue;
+      const { message } = item;
+      if (message.id !== finalAssistantId || message.role !== 'assistant') {
+        continue;
+      }
+      const finalAnswerDetail = compactTimelineText(message.content, 180, {
+        stripMarkdown: true,
+      });
+      if (finalAnswerDetail) return finalAnswerDetail;
```
Comment (wenshao) [inline packages/web-shell/client/components/MessageList.tsx:679]: **[Critical]** Content gap when stripped final answer is empty. `findFinalAnswerIndex` uses `isAssistantAnswer` which checks raw `content.trim().length > 0`, but `compactTimelineText(message.content, 180, { stripMarkdown: true })` may produce an empty string from non-empty raw content (e.g. `"> "`, `""`). When that happens, this early-return correctly skips, but the snippet loop below (line ~649) also skips the same message via `finalAssistantId`. Neither path covers it — the turn detail degrades to kind labels or "No activity".

**Fix:** Align the selection with the rendering — either make `isAssistantAnswer` use the stripped-content check, or don't skip the final answer in the snippet loop when the early-return produced empty.

_— qwen3.7-max via Qwen Code /review_

File: packages/web-shell/client/components/MessageList.test.ts [hunk 8]
```diff
@@ -563,6 +594,25 @@ describe('getSessionTimelineEntries', () => {
...
+    ]);
+
+    expect(entry?.label).toBe('介绍下 `agent-reproduce-align`');
+    expect(entry?.detail).toBe(
+      'agent-reproduce-align – 对齐测试技能 用途： 在 Qwen Code 中运行参考代码，中文强调·范围—引用「下划线，保留 snake_case。',
+    );
+  });
 });
 
 describe('getSessionTimelineRangeForIndexes', () => {
```
Comment (wenshao) [inline packages/web-shell/client/components/MessageList.test.ts:643]: **[Suggestion]** Five of the ten regex patterns in `cleanTimelineMarkdown` have no test coverage: images (`alt`), strikethrough (`~~text~~`), headings (`# Title`), blockquotes (`> text`), and list markers (`- item`). Each is a distinct regex with its own edge-case surface.

Consider adding a single test that exercises all untested patterns:

```ts
it('strips images, strikethrough, headings, blockquotes, and list markers', () => {
  const [entry] = getSessionTimelineEntries([
    makeUserMessage('u1'),
    {
      ...makeAssistantMessage('a1'),
      content: '![img](url)\n~~removed~~\n## Title\n> quoted\n- item one\n* item two',
    },
  ]);
  expect(entry?.detail).toBe(
    'img removed Title quoted item one item two',
  );
});
```

_— qwen3.7-max via Qwen Code /review_

**Review by wenshao — CHANGES_REQUESTED** [review]

File: packages/web-shell/client/components/MessageList.tsx (hunk 6 above)
Comment (callmeYe) [reply packages/web-shell/client/components/MessageList.tsx:573]: Fixed in `dc5026333`: `cleanTimelineMarkdown` now protects fenced/inline code before marker cleanup and replaces the backreference bold regex with `stripBalancedTimelineMarker`, which scans with `indexOf` and returns linearly on unmatched delimiters. Local validation: `cd packages/web-shell && npx vitest run client/components/MessageList.test.ts client/components/MessageList.dom.test.tsx` (114 passed) and `cd packages/web-shell && npm run build`.

File: packages/web-shell/client/components/MessageList.tsx (hunk 7 above)
Comment (callmeYe) [reply packages/web-shell/client/components/MessageList.tsx:679]: Fixed in `dc5026333`: `finalAssistantId` is selected only when the stripped final-answer preview is non-empty, so markdown-only final answers fall back to the normal turn detail path instead of suppressing it. Added a regression with `` that now falls back to the commentary label. Same focused tests and web-shell build pass locally.

File: packages/web-shell/client/components/MessageList.test.ts (hunk 8 above)
Comment (callmeYe) [reply packages/web-shell/client/components/MessageList.test.ts:643]: Covered in `dc5026333`: added one preview-cleaning regression that exercises image alt text, strikethrough, headings, blockquotes, list markers, fenced code, and inline code preservation. Focused MessageList tests now pass at 114/114, and web-shell build passes.

**Comment by wenshao** [comment]
## ✅ Maintainer verification — real local build & test (recommend merge)

I reproduced this PR against a fresh worktree at the PR head (`5fd635a`, merge of `main` into the branch) and drove it end-to-end: **unit + DOM tests, full-suite regression, production build, typecheck/lint, mutation testing, and a real Chromium render of the timeline** (jsdom can't measure layout). Net diff is exactly the 5 files described (166+/57−). Every claim in the PR description checks out.

### 1. Tests & build
| Check | Result |
|---|---|
| Full `web-shell` vitest suite | **628 / 628 pass** |
| Author's two files (`MessageList.test.ts` + `.dom.test.tsx`) | **112 pass** (88 + 24) |
| `npm run build` (vite ×2 + `tsc -p tsconfig.lib.json`) | **pass**, only the pre-existing Vite chunk-size warning |
| Production typecheck `tsc -p tsconfig.lib.json` (the CI gate) | **clean** |
| eslint on the 4 changed `.ts/.tsx` files | **clean** |

> Note: the PR body says 102 tests; the branch has since merged `main`, so the current head reports **112** in those two files — a superset, all green.

### 2. Mutation testing — the new tests are load-bearing
Reverted each PR logic change (keeping the PR's tests) and confirmed the mapped test fails. **6 / 6 caught:**

| Reverted change | Test that caught it |
|---|---|
| markdown stripping (`stripMarkdown`) | `cleans markdown markers from timeline details` |
| final-answer-preference early return | `uses the final answer detail without exposing thinking content` |
| "answer must follow the last work step" (`findFinalAnswerIndex`) | `does not prefer assistant text before later tool work` |
| `SESSION_TIMELINE_MIN_VISIBLE_ENTRIES = 4` | `hides the session timeline until there are at least four turns` |
| `hideSessionTimeline` gate | `hides the session timeline when the caller disables it` |
| native `title` removal | `renders the session timeline in the left gutter…` (`hasAttribute('title') === false`) |

### 3. Real Chromium render — the visual/CSS claims jsdom can't cover
Mounted the **real `MessageList` + real CSS module** in a headless Chromium harness (same 5 child-module stubs the DOM test uses) at a 1360px-wide container, and measured live:

| Description claim | Measured (live DOM) | ✓ |
|---|---|---|
| Rail appears at ≥4 assistant entries | `session-timeline` present, 4 entries | ✅ |
| Longest active/hovered tick ≈ **38px** | hovered tick = **38.0px** (10 × `scaleX(3.8)`) | ✅ |
| Shortest tick unchanged (compact base) | base ticks = **10px** each | ✅ |
| Rail sits closer to the message column | layer transform = `matrix(1,0,0,1,-22,0)` = **`translateX(-22px)`** | ✅ |
| Details card closer to the rail | computed `left` = **56px** | ✅ |
| No browser-native hover `title` | `anyButtonHasTitle: false` | ✅ |
| Sparse session (<4) hides the rail | 3 turns → rail **absent** | ✅ |
| Wide mode / caller-disable hides the rail | `hideSessionTimeline` → rail **absent** | ✅ |
| Cleaned summary matches the visible reply | see below | ✅ |

Rendered detail cards (markdown source → cleaned preview), confirming the summary logic **end-to-end in a browser**:
- **Final answer preferred + markdown stripped:** `**agent-reproduce-align** … [Qwen Code](url) … `` `ref` `` … *强调* … _下划线_, snake_case` → `agent-reproduce-align – 对齐测试技能 用途： 在 Qwen Code 中运行 ref 代码，中文强调·范围—引用「下划线，保留 snake_case。` (bold/link/code/emphasis removed, **`snake_case` preserved**).
- **Thinking never leaks:** turn with hidden reasoning → detail = `Plain answer two.`
- **Assistant text before later tool work is not the "final answer":** → detail = `I'll check · 1 tool call`.

**CSS A/B (proves the measurement tracks this PR's CSS):** reverting *only* `MessageList.module.css` to the merge-base yields the old values — base tick **12px**, hovered **52.8px** (12 × 4.4), transform **`none`**, details **`left: 72px`** — and restoring the PR flips them back to 10 / 38 / −22 / 56.

### 4. Reverse audit — no correctness or performance issues
- **No per-token recompute:** `getSessionTimelineSignature` uses a literal `'streaming'` sentinel for in-flight assistant/thinking messages, so `getSessionTimelineEntries` (and the new `cleanTimelineMarkdown`) recomputes **once per turn completion**, not on every streamed token.
- `cleanTimelineMarkdown` regexes are bounded (lazy quantifiers + negated classes — no ReDoS), ordered correctly (links before emphasis), and preserve `snake_case`; empty-after-clean content falls back gracefully to the snippet summary.
- `hideSessionTimeline={effectiveChatWidthMode === 'wide'}` is type-safe (`'wide'` is a valid `ChatWidthMode` literal) and only fires in explicit wide mode.

### Minor, non-blocking (optional)
Under `tsc -p tsconfig.json` (which includes test files and is **not** a CI gate — `web-shell` has no `typecheck` script; the build only checks `tsconfig.lib.json`), the new test fixtures add **+4** `DaemonMessage` union-spread type-looseness errors of the exact pattern already present on the base (5 → 9). No runtime or CI impact; could be tidied if desired.

**Verdict: correct, well-covered, and every described behavior is independently reproduced. 👍 Recommend merge.**

<sub>Verified on macOS (Darwin 25.5). Method: fresh worktree at PR head, full dep build (core→acp-bridge→sdk→webui→web-shell), vitest 628/628, 6 mutation A/Bs, real-Chromium harness + CSS A/B.</sub>

<details>
<summary>中文版（完整对应）</summary>

## ✅ 维护者验证 —— 本地真实构建与测试（建议合并）

我在 PR head（`5fd635a`，把 `main` 合入分支的 merge commit）的全新 worktree 上复现了本 PR，并做了端到端验证：**单元 + DOM 测试、全量回归、生产构建、typecheck/lint、变异测试，以及对 timeline 的真实 Chromium 渲染**（jsdom 无法测量真实布局）。净 diff 恰好是描述中的 5 个文件（166+/57−）。描述中的每一项声明都得到印证。

### 1. 测试与构建
| 检查 | 结果 |
|---|---|
| `web-shell` 全量 vitest | **628 / 628 通过** |
| 作者声明的两个文件（`MessageList.test.ts` + `.dom.test.tsx`） | **112 通过**（88 + 24） |
| `npm run build`（vite ×2 + `tsc -p tsconfig.lib.json`） | **通过**，仅有既有的 Vite chunk-size 警告 |
| 生产 typecheck `tsc -p tsconfig.lib.json`（真正的 CI 门禁） | **干净** |
| 对 4 个改动 `.ts/.tsx` 文件跑 eslint | **干净** |

> 说明：PR 描述写的是 102 个测试；分支之后合入了 `main`，当前 head 在这两个文件里是 **112**（是超集，全绿）。

### 2. 变异测试 —— 新测试是"承重"的
逐一把每处 PR 逻辑改动还原（保留 PR 自带的测试），确认对应测试会失败。**6 / 6 全部被捕获：**

| 被还原的改动 | 捕获它的测试 |
|---|---|
| markdown 清理（`stripMarkdown`） | `cleans markdown markers from timeline details` |
| "优先最终答案"的提前返回 | `uses the final answer detail without exposing thinking content` |
| "答案必须在最后一个工作步骤之后"（`findFinalAnswerIndex`） | `does not prefer assistant text before later tool work` |
| `SESSION_TIMELINE_MIN_VISIBLE_ENTRIES = 4` 阈值 | `hides the session timeline until there are at least four turns` |
| `hideSessionTimeline` 开关 | `hides the session timeline when the caller disables it` |
| 移除原生 `title` | `renders the session timeline in the left gutter…`（`hasAttribute('title') === false`） |

### 3. 真实 Chromium 渲染 —— jsdom 覆盖不到的视觉/CSS 声明
在无头 Chromium harness 中挂载**真实的 `MessageList` + 真实 CSS module**（用与 DOM 测试相同的 5 个子模块 stub），容器宽 1360px，实时测量：

| 描述中的声明 | 实测（真实 DOM） | ✓ |
|---|---|---|
| ≥4 个 assistant entry 时出现 rail | `session-timeline` 存在，4 条 entry | ✅ |
| active/hover 时最长横线 ≈ **38px** | hover 横线 = **38.0px**（10 × `scaleX(3.8)`） | ✅ |
| 最短横线保持不变（紧凑基准） | 基准横线各 **10px** | ✅ |
| rail 更靠近消息列 | layer transform = `matrix(1,0,0,1,-22,0)` = **`translateX(-22px)`** | ✅ |
| 详情卡片更靠近 rail | 计算后 `left` = **56px** | ✅ |
| 无浏览器原生 hover `title` | `anyButtonHasTitle: false` | ✅ |
| 稀疏会话（<4）隐藏 rail | 3 个 turn → rail **不渲染** | ✅ |
| 宽屏模式/调用方禁用时隐藏 rail | `hideSessionTimeline` → rail **不渲染** | ✅ |
| 摘要清理后与可见回复一致 | 见下 | ✅ |

渲染出的详情卡（markdown 源 → 清理后预览），在**浏览器里端到端**印证摘要逻辑：
- **优先最终答案 + 清理 markdown：** `**agent-reproduce-align** … [Qwen Code](url) … `` `ref` `` … *强调* … _下划线_, snake_case` → `agent-reproduce-align – 对齐测试技能 用途： 在 Qwen Code 中运行 ref 代码，中文强调·范围—引用「下划线，保留 snake_case。`（去掉了 加粗/链接/行内代码/强调，**保留了 `snake_case`**）。
- **thinking 不泄漏：** 含隐藏推理的 turn → 详情 = `Plain answer two.`
- **工具工作之前的 assistant 文本不算"最终答案"：** → 详情 = `I'll check · 1 tool call`。

**CSS A/B（证明测量确实跟随本 PR 的 CSS）：** 仅把 `MessageList.module.css` 还原到 merge-base 后得到旧值 —— 基准横线 **12px**、hover **52.8px**（12 × 4.4）、transform **`none`**、详情 **`left: 72px`**；恢复 PR 后又变回 10 / 38 / −22 / 56。

### 4. 反向审计 —— 无正确性或性能问题
- **不会每 token 重算：** `getSessionTimelineSignature` 对流式中的 assistant/thinking 消息用字面量 `'streaming'` 作签名，因此 `getSessionTimelineEntries`（以及新增的 `cleanTimelineMarkdown`）**每个 turn 完成时只算一次**，而非每个流式 token 都算。
- `cleanTimelineMarkdown` 的正则是有界的（惰性量词 + 取反字符类 —— 无 ReDoS），顺序正确（链接先于强调），并保留 `snake_case`；清理后为空时会优雅回退到 snippet 摘要。
- `hideSessionTimeline={effectiveChatWidthMode === 'wide'}` 类型安全（`'wide'` 是合法的 `ChatWidthMode` 字面量），只在显式宽屏模式触发。

### 次要、不阻塞（可选）
在 `tsc -p tsconfig.json`（包含测试文件，且**不是** CI 门禁 —— `web-shell` 没有 `typecheck` 脚本，构建只检查 `tsconfig.lib.json`）下，新测试 fixture 新增了 **+4** 个 `DaemonMessage` union-spread 的类型宽松报错，属于 base 上已存在的同一模式（5 → 9）。对运行时和 CI 都无影响；若愿意可顺手清理。

**结论：实现正确、覆盖充分、描述中的每一项行为都被独立复现。👍 建议合并。**

<sub>验证环境 macOS（Darwin 25.5）。方法：PR head 全新 worktree、完整依赖构建（core→acp-bridge→sdk→webui→web-shell）、vitest 628/628、6 个变异 A/B、真实 Chromium harness + CSS A/B。</sub>

</details>

**Review by wenshao — APPROVED** [review]

**Comment by wenshao** [comment]
@qwen-code /triage

**Review by DragonnZhang — APPROVED** [review]
Incremental review on new commit `8307c4c`: changes since last review look good. Timeline minimum-entries gate, markdown cleaning, CSS tuning, and `hideSessionTimeline` prop are all well-structured with thorough test coverage. CI all pass. No new issues found.

_— qwen3-coder via Qwen Code /review_
